src/client/clientlauncher.cpp

```diff
@@ -53,17 +53,13 @@ MainGameCallback *g_gamecallback = NULL;
 
 ClientLauncher::~ClientLauncher()
 {
-	if (receiver)
-		delete receiver;
+	delete receiver;
 
-	if (input)
-		delete input;
+	delete input;
 
-	if (g_fontengine)
-		delete g_fontengine;
+	delete g_fontengine;
 
-	if (device)
```

Comment:
The implementation of **drop()** is included in [IReferenceCounted.h](http://irrlicht.sourceforge.net/docu/_i_reference_counted_8h_source.html). Relevant code snippet:

```cpp
bool drop() const
{
    // someone is doing bad reference counting.
    _IRR_DEBUG_BREAK_IF(ReferenceCounter <= 0)
    --ReferenceCounter;
    if (!ReferenceCounter)
    {
        delete this;
        return true;
    }
    return false;
}
```

So it literally does standard C++ delete, plus updates the reference counter - unless something's wrong with the counter, then it returns false. The check in Minetest was redundant, but we could refactor the code to check for false, just to be on the safe side. Implementation-wise it would go like so:

```
    if (!device->drop())
        // something bad happened, trigger an error
```